### ISSUE TITLE
New Linking Workflow

### ISSUE THREAD

**Issue by guybedford** [jspm/jspm-cli#622]
We're strongly looking at this direction for linking scenarios:

```
  jspm link github:components/jquery@1.2.3 http://site.com/jquery.zip
  jspm link jquery ../some/folder
  jspm link jquery ../some/local/tarball.tar.gz
```

When linking a local folder, a direct symlink would be created. When linking a URL it would be downloaded and extracted.

The link command can be run again to update the installed package (eg updating dependencies, configuration etc).

We could also enable a `relink` command to relink based on where something is already linked:

```
  jspm relink jquery
```

Already knows that jquery was linked to `../some/local/tarball.tar.gz` so can regenerate from that same path.

**Comment by lookfirst** [comment]
What happens if the jquery in the .zip file is not 1.2.3?

**Comment by bfil** [comment]
So everyone using a project with archived dependencies will have to link them manually?
Since they are dependencies, shouldn't they work like any other, as in contained in the `package.json` file?

**Comment by guybedford** [comment]
@lookfirst it gets treated like version 1.2.3. Note you can leave the version or registry name out when linking and it is read from the package.json instead.

@bfil so here's the argument - URLs are a completely separate transport mechanism to a registry download, so we could consider having a custom registry endpoint just for this, ssee https://github.com/jspm/jspm-cli/issues/424. But then the issue becomes how to manage versioning.

The options seem to be:
1. Allow for registry endpoints which are unversioned.
2. Find other alternatives like this one here.

Thinking through the linking argument further how often would you say you use URLs in order to reference unpublished versions of something whose major version is already published on npm? Versus using URLs as the primary install target for something?

I suspect it may be the former, which is partly the inspiration for this.

**Comment by subesokun** [comment]
I think it would help if we define some scenarios where installing from archives makes sense. As you've already mentioned, why would somebody download and install a package from an URL or an archive when the package gets already hosted on npm or github?

Some scenarios:
* Package is very experimental/new and not yet hosted on a public hosting server. The effort for hosting this package on a public hosting server might not make sense if it's just a small dummy package for experimental purposes only
 - Probably linking in the archive into a version range might be the best solution for that

* You've a small kick-starter project with a few packages that you initially don't want to host public server
 - In that case you might be best of setting up a private git server and use a registry like jspm-git

* ...

**Comment by bfil** [comment]
My experience with it is the following: working on some private pet projects that go through a CI pipeline and get published only if the build and tests pass.

In addition, lots of my personal projects are contained in folders inside a single private GitHub repo (to avoid having to spend loads of money on GitHub), not ideal really and I know.

One way of doing it is publishing versioned archives on some file hosting at the end of the CI pipeline (like Amazon S3 for example) not ideal as well but works fine.

I worked around my issues by setting up my private NPM repository with [Sinopia](https://github.com/rlidwka/sinopia)

**Comment by subesokun** [comment]
I see, quite a complicated use case. So in the end if you don't want to host a private git or npm server you'd need something like 
```bash
// package.json
"dependencies": {
  "some-pkg": "http:site.com/some-pkg-1.2.3.zip"
}
```
though that would mean that you loose any support to define version ranges or doing a `jspm update` to fetch the latest versions of your package. Defining such kind of dependency is ok for private development (e.g. in a small team) but in case you want to officially publish your package this would mean that everybody who wants to install your package will depend on the availability of your `site.com`.

**Comment by bfil** [comment]
Exactly, as I said the use case is not ideal, but static file hosting is cheap and there's no need to run specific software like git or an npm server on them. Works ok for private projects when you don't want to spend money on hosting a repository server.

**Comment by guybedford** [comment]
We're strongly looking at this direction for linking scenarios:

```
  jspm link github:components/jquery@1.2.3 http://site.com/jquery.zip
  jspm link jquery ../some/folder
  jspm link jquery ../some/local/tarball.tar.gz
```

When linking a local folder, a direct symlink would be created. When linking a URL it would be downloaded and extracted.

The link command can be run again to update the installed package (eg updating dependencies, configuration etc).

We could also enable a `relink` command to relink based on where something is already linked:

```
  jspm relink jquery
```

Already knows that jquery was linked to `../some/local/tarball.tar.gz` so can regenerate from that same path.

**Comment by guybedford** [comment]
Updated description to match current directions.

**Comment by glen-84** [comment]
You could abstract it further and link from one registry/endpoint to another, like github -> npm, LOL.

1. Can the 2nd argument be any valid URI? f.e. ftp://... or ssh2://... ?

2. Is this command run from the linked package's folder or from the folder of the app that will use the linked package?

3. Where will the linking information be stored?

4. If folders are directly symlinked, will it still respect the files/directories/ignore options in the package.json file?

**Comment by guybedford** [comment]
@glen-84 good questions! Not sure about the meta linking though.

> 1. Can the 2nd argument be any valid URI? f.e. ftp://... or ssh2://... ?

We could certainly look to expand the URI schemas - initially it would just be file paths and URLs.

> 2. Is this command run from the linked package's folder or from the folder of the app that will use the linked package?

It is run in the app you are installing into.

> 3. Where will the linking information be stored?

We don't maintain any reference to the link source, apart from the fact that it is a symlink when doing file system links - you would run the link command equivalently again to update URL or other protocol links.

> 4. If folders are directly symlinked, will it still respect the files/directories/ignore options in the package.json file?

We just support the directories option for symlinking, not files and ignore.

**Comment by glen-84** [comment]
I think that it should be scheme-based, with a default (implied) scheme of "file:". It doesn't look like Node supports other schemes without additional packages though, so you may need some kind of extension mechanism.

Where will archives be extracted? (I'm trying to understand how the relink command knows where to find the original archive)

**Comment by guybedford** [comment]
@glen-84 the archives are extracted directly into `jspm_packages`. We could potentially store a `.linksource` file or something like that to know how to relink.

**Comment by glen-84** [comment]
Ah, I see.

Maybe .jspm or .jspm.json if you need to store additional metadata.

**Comment by guybedford** [comment]
Perhaps it's even worth considering here a `jspm install http://.../file.zip` basically just being a shorthand for this linking operation.

**Comment by subesokun** [comment]
It'd be awesome if its possible via this .linksource file to easily toggle the config.js between linked and none-linked mode.

**Comment by glen-84** [comment]
> Perhaps it's even worth considering here a `jspm install http://.../file.zip` basically just being a shorthand for this linking operation.

I assume you mean to extract the registry info from package.json in the ZIP archive. That's fine, but perhaps it's better to use a shorthand on the `link` command rather than on the `install` command, since the `http:` could be confused with a registry name.

**Comment by guybedford** [comment]
Effectively we would be making `http:`, `git:` etc etc special reserved names that can't be used as registry names, and then giving them special treatment.

**Comment by glen-84** [comment]
I don't like it, it makes it unclear what `x:` actually means.

You would then need to change the syntax for registries to use a different character, like `npm!x@y`, which I doubt that you want to do.

**Comment by peteruithoven** [comment]
`jspm link jquery ../some/folder` would create symlink directly to that folder (instead of a ~/.jspm/... folder)? 
That would be really really nice. That fixes the whole watch issue in #481. If this is possible this could really improve the workflow when working in multiple packages at the same time.

**Comment by michalstocki** [comment]
Any progress on this feature?

We need to install dependencies from maven artifactory, which are distributed in tarballs. We're considering writing custom JSPM Registry endpoint, however tarball dependencies may be a nice workaround. What is suggested solution for that issue?

**Comment by guybedford** [comment]
Linking archive files has been left out of the new 0.17 linking implementation leaving link to just work with symlinks. 

So the archive file install use case is still not covered, and a custom registry endpoint may well be advisable.

**Comment by guybedford** [comment]
Released in jspm 0.17 beta.
